File: ravenroot/ravenroot-ui/src/property-condition.js
// Ports `ai.ravenroot.api.catalog.PropertyCondition`/`PropertyConditionOperator`
// (ravenroot-application-api) to the editor. Read
// `docs/getting-started/conditional-node-properties.md` first — this file implements exactly that
// contract and adds nothing to it: no negation, no cross-node references, no expression combinators.
//
// ── The security rule this file exists to carry ─────────────────────────────────────────────────
//
// `PropertyCondition#CONTRACT` is versioned and travels with every serialized condition specifically
// so a consumer that does not recognise it can refuse to guess. A consumer that ignored an unknown
// contract and evaluated the condition anyway would silently turn a guarded field into an
// unconditional one — the failure is invisible, because the field just always appears, exactly as if
// no condition had been declared at all. `conditionHolds` below refuses first, before it ever looks
// at `operator` or `values`, so an unrecognised contract can never reach the switch that would
// otherwise evaluate it.
//
// The same function also carries the frontend's only defence-in-depth obligation: the backend
// validates every descriptor fail-closed at catalog load (bad references, cycles, inadmissible
// values), so a validated catalog can never actually SERVE a malformed condition — the Inspector is
// not a second validator and must not become one. But "cannot happen given a validated catalog" and
// "cannot arrive at this function" are different claims, and this module is written to the second,
// narrower one: any condition shape it does not fully recognise — unknown contract, unknown
// operator — is treated as unsatisfiable, never as holding. Fail-closed, not fail-silent.
//
// ── Why there is nothing here that looks like recursion ─────────────────────────────────────────
//
// A condition names one sibling PROPERTY and reads that sibling's current raw VALUE — never another
// property's computed visibility or required-ness. Evaluating property A's condition therefore never
// evaluates property B's condition, even when B's own condition happens to name A back: each lookup
// is an O(1) read of a value already sitting in the caller's `valuesByName` map. A descriptor that
// declared a genuine cycle (A depends on B depends on A) is rejected at catalog load on the backend
// and can never reach here — but even if a malformed one somehow did, there is no call graph in this
// module for it to be a cycle IN, which is a property of the design, not a check this file performs.

import { adapterIdOf } from './adapter-binding.js';

/** The one contract version this module understands. Mirrors `PropertyCondition.CONTRACT`. */
export const CONTRACT = 'ravenroot.property-condition/1';

/**
 * Whether `condition` holds against `siblingRawValue`, the sibling property's current (possibly
 * unsaved, possibly blank) value as the editor holds it right now.
 *
 * Mirrors `PropertyCondition#holds(String)` exactly, including its blank/whitespace normalisation —
 * reused from `./adapter-binding.js`'s `adapterIdOf` rather than reimplemented, because it is the
 * same question (`String#strip()`-equivalent blankness) `NodePropertyDescriptor#adapterIdOf` already
 * answers on the Java side, and CORE-07/ already establishes that this project keeps exactly one
 * definition of it per language, not one per caller.
 *
 * Returns `false` — never throws, never falls through to "holds" — for anything this module does not
 * fully recognise: an unrecognised `contract`, or an operator outside the four this module knows.
 * `condition` itself is expected non-null; a caller holding a possibly-absent condition decides what
 * absence means (see `isPropertyVisible`/`isPropertyRequiredNow` below) before calling this, but a
 * nullish `condition` reaching this function anyway is also treated as unsatisfiable rather than
 * throwing, for the same fail-closed reason.
 */
export function conditionHolds(condition, siblingRawValue) {
  if (!condition || condition.contract !== CONTRACT) return false;
  const value = adapterIdOf(siblingRawValue);
  switch (condition.operator) {
    case 'EQUALS':
    case 'ONE_OF':
      return Array.isArray(condition.values) && condition.values.includes(value);
    case 'PRESENT':
      return value !== '';
    case 'BLANK':
      return value === '';
    default:
      // The operator set is closed by design (see PropertyConditionOperator's own javadoc for why
      // there is deliberately no fifth case) — this is not a case this module expects to reach
      // against a validated catalog, but an operator it does not recognise gets the same refusal an
      // unrecognised contract gets, not a silent "true".
      return false;
  }
}

/**
 * Whether `property` should be shown, given every sibling's current raw value in `valuesByName`
 * (keyed by property name). `null`/`undefined` `visibleWhen` means no condition was declared, which
 * is unconditionally visible — not the same statement as "a condition that happens to hold" (see the
 * doc's own "null means no condition declared" line).
 */
export function isPropertyVisible(property, valuesByName) {
  const condition = property?.visibleWhen;
  if (!condition) return true;
  return conditionHolds(condition, valuesByName?.[condition.property]);
}

/**
 * Whether `property` is required RIGHT NOW.
 *
 * A declared `requiredWhen` REPLACES the static `required` flag as the answer, it does not gate on
 * top of it — the doc's own worked example (and its HTTP example, `"required":false` carrying a
 * `requiredWhen`) declares the base flag `false` precisely because the property is optional by
 * default and `requiredWhen` is what makes it conditionally mandatory. Requiring the base flag to
 * already be `true` before `requiredWhen` could do anything would make that documented example
 * impossible to express. `null`/absent `requiredWhen` falls back to the static flag unconditionally
 * — the ordinary, unconditional case does not change.
 *
 * Deliberately never consults `visibleWhen` here: the two are load-time-validated so that
 * `requiredWhen` holding implies `visibleWhen` holding, but that implication is a property of the
 * DATA, not something this function re-derives. A caller that also needs visibility calls
 * `isPropertyVisible` separately (`catalogPropertyFieldsHtml` does, and additionally never emits the
 * native `required` attribute while hidden regardless of what this function returns, as defence in
 * depth against a descriptor that violated that invariant).
 */
export function isPropertyRequiredNow(property, valuesByName) {
  const condition = property?.requiredWhen;
  if (condition) {
    return conditionHolds(condition, valuesByName?.[condition.property]);
  }
  return Boolean(property?.required);
}

File: ravenroot/ravenroot-application-api/src/main/java/ai/ravenroot/api/catalog/PropertyCondition.java
package ai.ravenroot.api.catalog;

import java.util.List;
import java.util.Objects;

/**
 * A condition on a <em>sibling</em> property of the same node type.
 *
 * <h2>Sibling-scoped, single-property, no expressions</h2>
 * <p>A condition names exactly one other property of the same {@link NodeTypeDescriptor} and one
 * {@link PropertyConditionOperator}. It cannot reach another node, another node type, or the
 * execution payload; it cannot compose sub-expressions. Several conditions on one property combine
 * with <b>AND</b>. Disjunction over one sibling is {@link PropertyConditionOperator#ONE_OF};
 * disjunction across different siblings is deliberately not expressible, because a combinator is the
 * first step toward the expression language this design exists to avoid.
 *
 * <h2>Versioned</h2>
 * <p>{@link #CONTRACT} is emitted with every serialized condition, following the precedent
 * {@code SyntheticProvenance} set. A consumer that does not recognise the contract string must treat
 * the condition as unsatisfiable rather than guess — an unknown condition is not an absent one.
 *
 * @param contract the versioned contract marker; see {@link #CONTRACT}
 * @param property the sibling property's {@code name}, within the same node type
 * @param operator the comparison; see {@link PropertyConditionOperator} for why the set is closed
 * @param values   operands. Empty for {@code PRESENT}/{@code BLANK}, exactly one for {@code EQUALS},
 *                 at least one for {@code ONE_OF}
 */
public record PropertyCondition(String contract, String property, PropertyConditionOperator operator,
                                List<String> values) {
    /** The versioned contract marker for this shape. */
    public static final String CONTRACT = "ravenroot.property-condition/1";

/**
 * Validates operand cardinality for the selected closed comparison operator.
 */
    public PropertyCondition {
        contract = contract == null || contract.isBlank() ? CONTRACT : contract;
        Objects.requireNonNull(operator, "operator");
        if (property == null || property.isBlank()) {
            throw new IllegalArgumentException("A property condition must name a sibling property");
        }
        values = values == null ? List.of() : List.copyOf(values);
        switch (operator) {
            case EQUALS -> {
                if (values.size() != 1) {
                    throw new IllegalArgumentException(
                            "EQUALS takes exactly one value, got " + values.size() + " for '" + property + "'");
                }
            }
            case ONE_OF -> {
                if (values.isEmpty()) {
                    throw new IllegalArgumentException("ONE_OF takes at least one value for '" + property + "'");
                }
            }
            case PRESENT, BLANK -> {
                if (!values.isEmpty()) {
                    throw new IllegalArgumentException(
                            operator + " takes no values, got " + values.size() + " for '" + property + "'");
                }
            }
        }
    }

/**
 * Creates an exact-value condition for one sibling property.
 * @param property sibling property name
 * @param value sole value that satisfies the condition
 * @return an {@link PropertyConditionOperator#EQUALS} condition
 */
    public static PropertyCondition equalTo(String property, String value) {
        return new PropertyCondition(CONTRACT, property, PropertyConditionOperator.EQUALS, List.of(value));
    }

/**
 * Creates a finite-set condition for one sibling property.
 * @param property sibling property name
 * @param values one or more permitted values
 * @return an {@link PropertyConditionOperator#ONE_OF} condition
 */
    public static PropertyCondition oneOf(String property, String... values) {
        return new PropertyCondition(CONTRACT, property, PropertyConditionOperator.ONE_OF, List.of(values));
    }

/**
 * Creates a condition requiring a non-blank sibling value.
 * @param property sibling property name
 * @return a {@link PropertyConditionOperator#PRESENT} condition
 */
    public static PropertyCondition present(String property) {
        return new PropertyCondition(CONTRACT, property, PropertyConditionOperator.PRESENT, List.of());
    }

/**
 * Creates a condition requiring an absent or blank sibling value.
 * @param property sibling property name
 * @return a {@link PropertyConditionOperator#BLANK} condition
 */
    public static PropertyCondition blank(String property) {
        return new PropertyCondition(CONTRACT, property, PropertyConditionOperator.BLANK, List.of());
    }

    /**
     * Whether this condition holds for the given sibling value.
     *
     * <p>{@code null} and blank are the same state here, matching {@code BehaviorPropertySchema}'s
     * own treatment of an absent property: a blank value is "not configured", not a value of "".
 * @param siblingValue current sibling value; {@code null} and blank are equivalent
 * @return whether this operator accepts the normalized sibling value
     */
    public boolean holds(String siblingValue) {
        String value = siblingValue == null ? "" : siblingValue.strip();
        return switch (operator) {
            case EQUALS, ONE_OF -> values.contains(value);
            case PRESENT -> !value.isEmpty();
            case BLANK -> value.isEmpty();
        };
    }

/**
 * The satisfying operand set, for load-time checking against the sibling's allowed values.
 * @return immutable operands used by equality and membership operators
 */
    public List<String> operands() {
        return values;
    }
}
